src/types/Stream.d.ts

```diff
@@ -0,0 +1,40 @@
+import Client from './Client'
+import { ExprArg } from './query'
+import { FaunaTime, Ref } from './values'
+
+export interface Subscription {
+  on: (type: SubscriptionEvents, callback: Function) => this
+  start: () => this
+  close: () => this
+  event: string
+  txnTS: number
+  data:
+    | StreamStartEventData
+    | StreamVersionEventData
+    | StreamErrorEventData
+    | StreamHistoryRewriteEventData
+}
+
+type StreamStartEventData = FaunaTime
+
+type StreamVersionEventData = {
+  ref: Ref
+  ts: number
+  action: string
+  new: object
+  old: object
+  diff: string
+}
+
+type StreamErrorEventData = {
+  code: number
+  description: string
+}
+
+type StreamHistoryRewriteEventData = {
+  ref: Ref
+  ts: number
+  action: string
+}
+
+type SubscriptionEvents = 'start' | 'version' | 'history_rewrite' | 'error'
```

Comment:
Question: is it possible in TS to tie the literal string with the event type received?

Reply:
Found this. https://github.com/bterlson/strict-event-emitter-types